### DISCUSSION TITLE
【Module Federation】Error loading module - Cannot read property 'call' of undefined

### DISCUSSION THREAD

**Discussion by YD-Feng** [webpack/webpack#14037] [General]
# Bug report
**What is the current behavior?**
1.【dbcWebMain】app exposes some modules
2. webpack.config.js set optimization.runtimeChunk: 'single' for HMR work
3. HMR work well，but 【dbcWebSub】app Error loading module
4. if webpack.config.js set optimization.runtimeChunk: false，modules load success，but HMR not work

repository：
https://gitee.com/caozhifeng1000/webpack-5-demo.git

start app：
`npm run dev`

【dbcWebMain】app run at http://127.0.0.1:3927
【dbcWebSub】app run at http://127.0.0.1:4936

**Other relevant information:**
webpack version:  5.51.1
webpack-cli version: 4.8.0
webpack-dev-server version: 3.11.2
Node.js version: 14.16.1
Operating System: Win10

[a comment by an automated account is not shown]

**Comment by YD-Feng** [comment]
`
/******/ 		// Execute the module function
/******/ 		try {
/******/ 			var execOptions = { id: moduleId, module: module, factory: __webpack_modules__[moduleId], require: __webpack_require__ };
/******/ 			__webpack_require__.i.forEach(function(handler) { handler(execOptions); });
/******/ 			module = execOptions.module;
/******/ 			execOptions.factory.call(module.exports, module, module.exports, execOptions.require);
/******/ 		} catch(e) {
/******/ 			module.error = e;
/******/ 			throw e;
/******/ 		}
`

execOptions.factory is undefined

**Comment by franckXu** [comment]
https://webpack.js.org/blog/2020-12-08-roadmap-2021/#hot-module-replacement-for-module-federation

**Comment by alexander-akait** [comment]
Change `name` in `package.json` on different names, it was answered a lot of time

**Reply by alexander-akait** [reply]
And we already work on our docs about it

**Reply by deepsweet** [reply]
Could you please elaborate? I have the same error appearing if I import a federated module statically instead of dynamically.

**Reply by GVLevi** [reply]
If it was "answered a lot of times" then surly there's a place with documentation you can link to. A vague comment doesn't provide any insight as to what the underlying issue is or why it's occurring - so for those where the `name` differs and changing it in package.json doesn't fix it your comment provides no better context for individual resolution.

**Comment by JonWallsten** [comment]
@alexander-akait This issue started appearing for us after updating from:
```
"webpack": "5.64.2",
"webpack-cli": "4.9.1",
"webpack-dev-server": "4.5.0"
```
to: 
```
"webpack": "5.67.0",
"webpack-cli": "4.9.2",
"webpack-dev-server": "4.7.3"
```

Have you guys done any recent changes related to this?
![image](https://user-images.githubusercontent.com/5993909/151788484-fad50fe4-24b8-4065-b64e-ed80390890fa.png)

It's trying to load module 78.
But it doesn't exist:
![image](https://user-images.githubusercontent.com/5993909/151788566-8b96723f-76fb-4036-a0ac-0f9e3c14e4aa.png)

It does however exist in the cache:
![image](https://user-images.githubusercontent.com/5993909/151789262-96b34184-cac7-4dee-b3a9-8ac119a8d8f3.png)


I'll work on a repro if necessary. Just wanted to give you a quick heads up!

It only appears in one of our 5 applications with nearly identical setup.

**Reply by alexander-akait** [reply]
I think problem in circular deps

**Reply by alexander-akait** [reply]
But need investigate

**Reply by JonWallsten** [reply]
Just made an update. It exists in the cache.
I'll try to work on a repro than and check for circular deps.

**Reply by JonWallsten** [reply]
![image](https://user-images.githubusercontent.com/5993909/151789884-c3e59e62-3197-4ae6-9ae0-5db598a4dfe7.png)
Parent: 22 is the entry point "main.ts".

**Reply by JonWallsten** [reply]
@alexander-akait What I don't understand is, why would it appear in the cache but not the current list of modules? If there was a circular dependency, would it not fail before it's added to the cache as well? I have to point out that the other applications are using the same libraries without any issues. But they might not import the same things from the libs.

**Reply by JonWallsten** [reply]
I think I found the correct one now:
![image](https://user-images.githubusercontent.com/5993909/151796406-7cf82a56-05d8-4aae-9dab-15dfe3bab36b.png)
![image](https://user-images.githubusercontent.com/5993909/151796409-88942223-f109-4b16-a21a-a04df952b6da.png)

**Reply by JonWallsten** [reply]
It has the correct id and the correct parent id.

**Reply by JonWallsten** [reply]
It works fine when I remove the import.

**Reply by JonWallsten** [reply]
Same issue with 
![image](https://user-images.githubusercontent.com/5993909/151797141-fb2a84d2-cf0a-4033-af67-867a94ce8dcd.png)

**Reply by vankop** [reply]
could you please create a reproducible repo?

**Reply by JonWallsten** [reply]
Done:
https://github.com/JonWallsten/monorepo-new/tree/webpack-bug
**Install**
npm run full-install
**Run**
npm run dev

Go to http://localhost:4000

**Reply by vankop** [reply]
oh.. it is so huge.. could you try to detect which major webpack version affect this? also did not find how this relative to module federation? (or I did not find relative configuration)

**Reply by JonWallsten** [reply]
Yeah, it's the skeleton for our monorepo. They only way to find out what version is to try all of them I guess. I just left work. So it would have to be tomorrow. I'm not sure what module federation is. It's probably just the same error message in the end. Maybe it something related to the cache?

**Reply by vankop** [reply]
> I'm not sure what module federation is.

Ok, so probably error comes from on of apps in `~/packages`, this could be reduced to one webpack.config then..

**Reply by JonWallsten** [reply]
Yeah, its something in web-app-prime (the other apps works fine with the same import and basically the same config). But prime is using the libraries. But it this case since it's a direct import in web-app-prime from node_modules (normalize.css) we can probably rule out the libs and the other apps. But I'm really confused why only one of our apps is at issue here.

**Reply by JonWallsten** [reply]
@vankop It stopped working in 5.66.0. I removed that other apps and have the same issue. I have not removed the lib since the app is dependent on them.

**Reply by JonWallsten** [reply]
![image](https://user-images.githubusercontent.com/5993909/151927267-f5fd352c-1872-4450-9092-9f6bd1218b78.png)
![image](https://user-images.githubusercontent.com/5993909/151927274-387208ca-6bbf-432a-ba10-e5963fd6eee3.png)

**Reply by JonWallsten** [reply]
Ah, I see now. This was a mistake on my part. I have commented out `futureDefaults: true` everywhere but in prime, where I forgot to do it.

**Reply by vankop** [reply]
Could you create an issue with bug?

**Reply by JonWallsten** [reply]
If it is a bug? I might be because I'm using `mini-css-extract-plugin` and `css-loader` which this experiment replaces?

**Reply by vankop** [reply]
@alexander-akait should this be treated as bug?

**Reply by alexander-akait** [reply]
Need to look at code, maybe it is async loading CSS in initial chunk

**Reply by JonWallsten** [reply]
@alexander-akait You still have access to the repo since way back. I merged the branch to master to get the new dependencies. So if you want to look at the code you can check the master branch.

**Reply by corradin** [reply]
My issue was that I was missing a dynamic import. I renamed the existing index.js (I had an existing repo) to bootstrap.js, added a new index.js file with the line ```import('./bootstrap');```

**Reply by vedikak** [reply]
Hi 
I am facing the same challenge, is a conclusive solution reached? I havent been able to fix the execOptions error yet
